Library/brew.rb

```diff
@@ -39,13 +38,24 @@
   EOS
 end
 
+# Check for user agreement of the Xcode license before permitting
+# any other brew usage to continue. This prevents the situation where
+# people are instructed to "please re-run as root via sudo" on brew commands.
+# The check can only fail when Xcode is installed & the active developer dir.
+if `/usr/bin/xcrun clang 2>&1` =~ /license/ && !$?.success?
+  odie <<-EOS.undent
+    You have not agreed to the Xcode license. Please resolve this by running:
+      sudo xcodebuild -license
+  EOS
+end
```

Comment:
`update` and `tap` may relate to git. We can add some check in https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/utils/git.rb#L9

`install` and `upgrade` should belong to `doctor` check https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/cmd/install.rb#L135-L147

For `doctor` check, we can bypass some check if license check fails.